README.rst
NumRep
------

.. image:: https://travis-ci.org/necessary129/NumRep.svg
    :target: https://travis-ci.org/necessary129/NumRep


.. image:: https://coveralls.io/repos/necessary129/NumRep/badge.svg?branch=master&service=github
  :target: https://coveralls.io/github/necessary129/NumRep?branch=master

Version: 0.1.3

This module provides a way to represent the place values in a given number, eg:
    >>> a = NumRep(1234456789123)
    >>> a
    NumRep(Crores=123445,Lakhs=67,Thousands=89,Hundreds=1,Tens=2,Ones=3)
    >>> a.crores
    123445
    >>> a.ones
    3
    >>> a.hundreds
    1
    >>> a.tens
    2
    >>> a.lakhs
    67
    >>> print(a)
    123445 Crores, 67 Lakhs, 89 Thousands, 1 Hundred, 2 Tens, 3 Ones

This raises TypeError when you give it an invalid number, eg:
    >>> a = NumRep('notanumber1111')
    Traceback (most recent call last):
    ...
    TypeError: Not a valid Number.

And Every integer in the NumRep has a GetRep() method, which returns the Representation of the number itself, eg:
    >>> a = NumRep(123456789)
    >>> a
    NumRep(Crores=12,Lakhs=34,Thousands=56,Hundreds=7,Tens=8,Ones=9)
    >>> a.crores
    12
    >>> a.crores.GetRep()
    NumRep(Tens=1,Ones=2)
    >>> a.crores.GetRep().tens.GetRep()
    NumRep(Ones=1)

You can also do arithmetic operations on them and then use GetRep(), eg:
    >>> a = NumRep(123456789)
    >>> a
    NumRep(Crores=12,Lakhs=34,Thousands=56,Hundreds=7,Tens=8,Ones=9)
    >>> (a.crores - 2).GetRep()
    NumRep(Tens=1)

You can compare and do arithmetic operations NumReps with integers and also with other NumReps, like:
    >>> c = NumRep(15)
    >>> c
    NumRep(Tens=1,Ones=5)
    >>> c + 10
    NumRep(Tens=2,Ones=5)
    >>> b = NumRep(10)
    >>> b
    NumRep(Tens=1)
    >>> c + b
    NumRep(Tens=2,Ones=5)


This class also provides an all() function, which gives you the real value of the given denomination, eg:
    >>> a = NumRep(1234456789123)
    >>> a.all('hundreds')
    12344567891
You can also use the values specified in the module for all(), eg:
    >>> a = NumRep(1234456789123)
    >>> a.all(CRORES)
    123445

CHANGELOG.rst
ChangeLog
---------

0.1.3
 * Another bug fix

0.1.2
 * Fixed a major bug.

0.1.1
 * define a __str__() function
 * Add a __main__ module

0.1
 * Now can compare and do arithmetic operations on NumRep objects
 * Made NumRep a package

0.0.1.1
 * Added variabes for all() functions in the module

0.0.1
 * First Release

test.py
#!/usr/bin/env python

# Copyright (c) 2015 Muhammed Shamil K
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
# The above copyright notice and this permission notice shall be included in
#  all copies or substantial portions of the Software.
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
from __future__ import print_function

import sys

def testmod(mod):
    import doctest
    mod = __import__(mod)
    err, test = doctest.testmod(m=mod,verbose=True)
    if err > 0:
        return False
    else:
        return True

if __name__ == '__main__':
    import os
    first = testmod('NumRep')
    if not first:
        sys.exit(1)
    for mod in os.listdir('NumRep'):
        if mod.startswith('__') or mod.endswith('.pyc'):
            continue
        module = 'NumRep.'+mod.replace('.py','')
        test = testmod(module)
        if not test:
            sys.exit(1)


do_test.py
from test import support as s
from tests.testv2 import Tester
from tests.test_rst import runt

def testmod(mod):
    import doctest
    mod = __import__(mod)
    err, test = doctest.testmod(m=mod,verbose=True)
    if err > 0:
        return False
    else:
        return True

def main():
    import os
    first = testmod('NumRep')
    if not first:
        sys.exit(1)
    for mod in os.listdir('NumRep'):
        if mod.startswith('__') or mod.endswith('.pyc'):
            continue
        module = 'NumRep.'+mod.replace('.py','')
        test = testmod(module)
        if not test:
            sys.exit(1)

if __name__ == '__main__':
    main()
    s.run_unittest(Tester)
    runt('README.rst')
    runt('CHANGELOG.rst')

tests/testv2.py
#!/usr/bin/env python

# Copyright (c) 2015 Muhammed Shamil K
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
# The above copyright notice and this permission notice shall be included in
#  all copies or substantial portions of the Software.
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.

import unittest
from NumRep import NumRep, HUNDREDS, THOUSANDS, CRORES, TENS, ONES, LAKHS
from NumRep.funcs import RepNum

class Tester(unittest.TestCase):
    def setUp(self):
        self.numrep = NumRep(123456789111)

    def test_repr(self):
        self.assertEqual(repr(self.numrep),'NumRep(Crores=12345,Lakhs=67,Thousands=89,Hundreds=1,Tens=1,Ones=1)')

    def test_str(self):
        self.assertEqual(str(self.numrep),'12345 Crores, 67 Lakhs, 89 Thousands, 1 Hundred, 1 Ten, 1 One')

    def test_adds_int(self):
        calcd = self.numrep + 1111111111111
        self.assertIsInstance(calcd, NumRep)
        self.assertEqual(calcd, int((123456789111+1111111111111)))
        self.assertEqual(calcd.crores, 123456)
        self.assertEqual(calcd.lakhs, 79)
        self.assertEqual(calcd.thousands, 0)
        self.assertEqual(calcd.hundreds, 2)
        self.assertEqual(calcd.tens, 2)
        self.assertEqual(calcd.ones, 2)
        self.assertEqual(calcd.number, int((123456789111+1111111111111)))
        self.assertEqual(repr(calcd),'NumRep(Crores=123456,Lakhs=79,Hundreds=2,Tens=2,Ones=2)')
        self.assertEqual(str(calcd),'123456 Crores, 79 Lakhs, 2 Hundreds, 2 Tens, 2 Ones')

    def test_subtracts_int(self):
        calcd = self.numrep - 111111111111
        self.assertIsInstance(calcd, NumRep)
        self.assertEqual(calcd, int((123456789111-111111111111)))
        self.assertEqual(calcd.crores, 1234)
        self.assertEqual(calcd.lakhs, 56)
        self.assertEqual(calcd.thousands, 78)
        self.assertEqual(calcd.hundreds, 0)
        self.assertEqual(calcd.tens, 0)
        self.assertEqual(calcd.ones, 0)
        self.assertEqual(calcd.number, int((123456789111-111111111111)))
        self.assertEqual(repr(calcd),'NumRep(Crores=1234,Lakhs=56,Thousands=78)')
        self.assertEqual(str(calcd),'1234 Crores, 56 Lakhs, 78 Thousands')

    def test_multiplies_int(self):
        calcd = self.numrep * 2
        self.assertIsInstance(calcd, NumRep)
        self.assertEqual(calcd, int((123456789111*2)))
        self.assertEqual(calcd.crores, 24691)
        self.assertEqual(calcd.lakhs, 35)
        self.assertEqual(calcd.thousands, 78)
        self.assertEqual(calcd.hundreds, 2)
        self.assertEqual(calcd.tens, 2)
        self.assertEqual(calcd.ones, 2)
        self.assertEqual(calcd.number, int((123456789111*2)))
        self.assertEqual(repr(calcd),'NumRep(Crores=24691,Lakhs=35,Thousands=78,Hundreds=2,Tens=2,Ones=2)')
        self.assertEqual(str(calcd),'24691 Crores, 35 Lakhs, 78 Thousands, 2 Hundreds, 2 Tens, 2 Ones')

    def test_divs_int(self):
        calcd = self.numrep / 2
        self.assertIsInstance(calcd, NumRep)
        self.assertEqual(calcd, int((123456789111/2)))
        self.assertEqual(calcd.crores, 6172)
        self.assertEqual(calcd.lakhs, 83)
        self.assertEqual(calcd.thousands, 94)
        self.assertEqual(calcd.hundreds, 5)
        self.assertEqual(calcd.tens, 5)
        self.assertEqual(calcd.ones, 5)
        self.assertEqual(calcd.number, int((123456789111/2)))
        self.assertEqual(repr(calcd),'NumRep(Crores=6172,Lakhs=83,Thousands=94,Hundreds=5,Tens=5,Ones=5)')
        self.assertEqual(str(calcd),'6172 Crores, 83 Lakhs, 94 Thousands, 5 Hundreds, 5 Tens, 5 Ones')

    def test_mod_int(self):
        calcd = self.numrep % 22222222
        self.assertIsInstance(calcd, NumRep)
        self.assertEqual(calcd, int((123456789111%22222222)))
        self.assertEqual(calcd.crores, 1)
        self.assertEqual(calcd.lakhs, 23)
        self.assertEqual(calcd.thousands, 45)
        self.assertEqual(calcd.hundreds, 9)
        self.assertEqual(calcd.tens, 0)
        self.assertEqual(calcd.ones, 1)
        self.assertEqual(calcd.number, int((123456789111%22222222)))
        self.assertEqual(repr(calcd),'NumRep(Crores=1,Lakhs=23,Thousands=45,Hundreds=9,Ones=1)')
        self.assertEqual(str(calcd),'1 Crore, 23 Lakhs, 45 Thousands, 9 Hundreds, 1 One')

    def test_divmod_int(self):
        calcd, calc2d = divmod(self.numrep, 22222222)
        self.assertIsInstance(calcd, NumRep)
        self.assertIsInstance(calc2d, NumRep)
        self.assertEqual((calcd,calc2d), divmod(123456789111, 22222222))
        self.assertEqual(calcd.crores, 0)
        self.assertEqual(calcd.lakhs, 0)
        self.assertEqual(calcd.thousands, 5)
        self.assertEqual(calcd.hundreds, 5)
        self.assertEqual(calcd.tens, 5)
        self.assertEqual(calcd.ones, 5)
        self.assertEqual(calcd.number, int((123456789111//22222222)))
        self.assertEqual(repr(calcd),'NumRep(Thousands=5,Hundreds=5,Tens=5,Ones=5)')
        self.assertEqual(str(calcd),'5 Thousands, 5 Hundreds, 5 Tens, 5 Ones')
        self.assertEqual(calc2d.crores, 1)
        self.assertEqual(calc2d.lakhs, 23)
        self.assertEqual(calc2d.thousands, 45)
        self.assertEqual(calc2d.hundreds, 9)
        self.assertEqual(calc2d.tens, 0)
        self.assertEqual(calc2d.ones, 1)
        self.assertEqual(calc2d.number, int((123456789111%22222222)))
        self.assertEqual(repr(calc2d),'NumRep(Crores=1,Lakhs=23,Thousands=45,Hundreds=9,Ones=1)')
        self.assertEqual(str(calc2d),'1 Crore, 23 Lakhs, 45 Thousands, 9 Hundreds, 1 One')

    def test_adds_numrep(self):
        calcd = self.numrep + NumRep(1111111111111)
        self.assertIsInstance(calcd, NumRep)
        self.assertEqual(calcd, int((123456789111+1111111111111)))
        self.assertEqual(calcd.crores, 123456)
        self.assertEqual(calcd.lakhs, 79)
        self.assertEqual(calcd.thousands, 0)
        self.assertEqual(calcd.hundreds, 2)
        self.assertEqual(calcd.tens, 2)
        self.assertEqual(calcd.ones, 2)
        self.assertEqual(calcd.number, int((123456789111+1111111111111)))
        self.assertEqual(repr(calcd),'NumRep(Crores=123456,Lakhs=79,Hundreds=2,Tens=2,Ones=2)')
        self.assertEqual(str(calcd),'123456 Crores, 79 Lakhs, 2 Hundreds, 2 Tens, 2 Ones')

    def test_subtracts_numrep(self):
        calcd = self.numrep - NumRep(111111111111)
        self.assertIsInstance(calcd, NumRep)
        self.assertEqual(calcd, int((123456789111-111111111111)))
        self.assertEqual(calcd.crores, 1234)
        self.assertEqual(calcd.lakhs, 56)
        self.assertEqual(calcd.thousands, 78)
        self.assertEqual(calcd.hundreds, 0)
        self.assertEqual(calcd.tens, 0)
        self.assertEqual(calcd.ones, 0)
        self.assertEqual(calcd.number, int((123456789111-111111111111)))
        self.assertEqual(repr(calcd),'NumRep(Crores=1234,Lakhs=56,Thousands=78)')
        self.assertEqual(str(calcd),'1234 Crores, 56 Lakhs, 78 Thousands')

    def test_multiplies_numrep(self):
        calcd = self.numrep * NumRep(2)
        self.assertIsInstance(calcd, NumRep)
        self.assertEqual(calcd, int((123456789111*2)))
        self.assertEqual(calcd.crores, 24691)
        self.assertEqual(calcd.lakhs, 35)
        self.assertEqual(calcd.thousands, 78)
        self.assertEqual(calcd.hundreds, 2)
        self.assertEqual(calcd.tens, 2)
        self.assertEqual(calcd.ones, 2)
        self.assertEqual(calcd.number, int((123456789111*2)))
        self.assertEqual(repr(calcd),'NumRep(Crores=24691,Lakhs=35,Thousands=78,Hundreds=2,Tens=2,Ones=2)')
        self.assertEqual(str(calcd),'24691 Crores, 35 Lakhs, 78 Thousands, 2 Hundreds, 2 Tens, 2 Ones')

    def test_divs_numrep(self):
        calcd = self.numrep / NumRep(2)
        self.assertIsInstance(calcd, NumRep)
        self.assertEqual(calcd, int((123456789111/2)))
        self.assertEqual(calcd.crores, 6172)
        self.assertEqual(calcd.lakhs, 83)
        self.assertEqual(calcd.thousands, 94)
        self.assertEqual(calcd.hundreds, 5)
        self.assertEqual(calcd.tens, 5)
        self.assertEqual(calcd.ones, 5)
        self.assertEqual(calcd.number, int((123456789111/2)))
        self.assertEqual(repr(calcd),'NumRep(Crores=6172,Lakhs=83,Thousands=94,Hundreds=5,Tens=5,Ones=5)')
        self.assertEqual(str(calcd),'6172 Crores, 83 Lakhs, 94 Thousands, 5 Hundreds, 5 Tens, 5 Ones')

    def test_mod_numrep(self):
        calcd = self.numrep % NumRep(22222222)
        self.assertIsInstance(calcd, NumRep)
        self.assertEqual(calcd, int((123456789111%22222222)))
        self.assertEqual(calcd.crores, 1)
        self.assertEqual(calcd.lakhs, 23)
        self.assertEqual(calcd.thousands, 45)
        self.assertEqual(calcd.hundreds, 9)
        self.assertEqual(calcd.tens, 0)
        self.assertEqual(calcd.ones, 1)
        self.assertEqual(calcd.number, int((123456789111%22222222)))
        self.assertEqual(repr(calcd),'NumRep(Crores=1,Lakhs=23,Thousands=45,Hundreds=9,Ones=1)')
        self.assertEqual(str(calcd),'1 Crore, 23 Lakhs, 45 Thousands, 9 Hundreds, 1 One')

    def test_divmod_numrep(self):
        calcd, calc2d = divmod(self.numrep, NumRep(22222222))
        self.assertIsInstance(calcd, NumRep)
        self.assertIsInstance(calc2d, NumRep)
        self.assertEqual((calcd,calc2d), divmod(123456789111, 22222222))
        self.assertEqual(calcd.crores, 0)
        self.assertEqual(calcd.lakhs, 0)
        self.assertEqual(calcd.thousands, 5)
        self.assertEqual(calcd.hundreds, 5)
        self.assertEqual(calcd.tens, 5)
        self.assertEqual(calcd.ones, 5)
        self.assertEqual(calcd.number, int((123456789111//22222222)))
        self.assertEqual(repr(calcd),'NumRep(Thousands=5,Hundreds=5,Tens=5,Ones=5)')
        self.assertEqual(str(calcd),'5 Thousands, 5 Hundreds, 5 Tens, 5 Ones')
        self.assertEqual(calc2d.crores, 1)
        self.assertEqual(calc2d.lakhs, 23)
        self.assertEqual(calc2d.thousands, 45)
        self.assertEqual(calc2d.hundreds, 9)
        self.assertEqual(calc2d.tens, 0)
        self.assertEqual(calc2d.ones, 1)
        self.assertEqual(calc2d.number, int((123456789111%22222222)))
        self.assertEqual(repr(calc2d),'NumRep(Crores=1,Lakhs=23,Thousands=45,Hundreds=9,Ones=1)')
        self.assertEqual(str(calc2d),'1 Crore, 23 Lakhs, 45 Thousands, 9 Hundreds, 1 One')

    def test_bool(self):
        a = NumRep(0)
        self.assertFalse(a)
        a = NumRep(1)
        self.assertTrue(a)

    def test_abs(self):
        self.assertEqual(abs(self.numrep), abs(123456789111))

    def test_lt_int(self):
        self.assertTrue(self.numrep < 123456789112)
        self.assertFalse(self.numrep < 123456789110)

    def test_gt_int(self):
        self.assertTrue(self.numrep > 123456789110)
        self.assertFalse(self.numrep > 123456789112)

    def test_ge_int(self):
        self.assertTrue(self.numrep >= 123456789110)
        self.assertFalse(self.numrep >= 123456789112)
        self.assertTrue(self.numrep >= 123456789111)

    def test_le_int(self):
        self.assertTrue(self.numrep <= 123456789112)
        self.assertFalse(self.numrep <= 123456789110)
        self.assertTrue(self.numrep <= 123456789111)

    def test_lt_numrep(self):
        self.assertTrue(self.numrep < NumRep(123456789112))
        self.assertFalse(self.numrep < NumRep(123456789110))

    def test_gt_numrep(self):
        self.assertTrue(self.numrep > NumRep(123456789110))
        self.assertFalse(self.numrep > NumRep(123456789112))

    def test_ge_numrep(self):
        self.assertTrue(self.numrep >= NumRep(123456789110))
        self.assertFalse(self.numrep >= NumRep(123456789112))
        self.assertTrue(self.numrep >= NumRep(123456789111))

    def test_le_numrep(self):
        self.assertTrue(self.numrep <= NumRep(123456789112))
        self.assertFalse(self.numrep <= NumRep(123456789110))
        self.assertTrue(self.numrep <= NumRep(123456789111))

    def test_all_func(self):
        self.assertEqual(self.numrep.all(CRORES),12345)
        self.assertEqual(self.numrep.all(LAKHS),1234567)
        self.assertEqual(self.numrep.all(THOUSANDS),123456789)
        self.assertEqual(self.numrep.all(HUNDREDS),1234567891)
        self.assertEqual(self.numrep.all(TENS),12345678911)
        self.assertEqual(self.numrep.all(ONES),123456789111)
        self.assertEqual(self.numrep.all('crores'),12345)
        self.assertEqual(self.numrep.all('lakhs'),1234567)
        self.assertEqual(self.numrep.all('thousands'),123456789)
        self.assertEqual(self.numrep.all('hundreds'),1234567891)
        self.assertEqual(self.numrep.all('tens'),12345678911)
        self.assertEqual(self.numrep.all('ones'),123456789111)

    def test_raises(self):
        with self.assertRaisesRegex(TypeError,'Not a valid Number'):
            NumRep('11a')
        with self.assertRaisesRegex(TypeError,'Not a valid denomination'):
            self.numrep.all('invalid')

    def test_repnum(self):
        self.assertIsInstance(self.numrep.crores,RepNum)
        calcd = self.numrep.crores.GetRep()
        self.assertIsInstance(calcd,NumRep)
        self.assertEqual(calcd, 12345)
        self.assertEqual(calcd.crores, 0)
        self.assertEqual(calcd.lakhs, 0)
        self.assertEqual(calcd.thousands, 12)
        self.assertEqual(calcd.hundreds, 3)
        self.assertEqual(calcd.tens, 4)
        self.assertEqual(calcd.ones, 5)
        self.assertEqual(calcd.number, 12345)
        self.assertEqual(repr(calcd),'NumRep(Thousands=12,Hundreds=3,Tens=4,Ones=5)')
        self.assertEqual(str(calcd),'12 Thousands, 3 Hundreds, 4 Tens, 5 Ones')


if __name__ == '__main__':
    unittest.main()


verchg.py
#!/usr/bin/env python

# Copyright (c) 2015 Muhammed Shamil K
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
# The above copyright notice and this permission notice shall be included in
#  all copies or substantial portions of the Software.
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.

import setup
vre = setup.vre

def write_version(fl,vrs):
    for x,y in enumerate(vrs):
        vrs[x] = str(y)
    vers = ".".join(vrs)
    version = "Version: {0}".format(vers)
    fl = vre.sub(version,fl)
    return fl

def do():
    readme = setup.read_file('README.rst')
    version = setup.getversion()
    version[-1] += 1
    new = write_version(readme,version)
    with open('README.rst','w') as f:
        f.write(new)

if __name__ == '__main__':
    do()

NumRep/utils.py
# Copyright (c) 2015 Muhammed Shamil K
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
# The above copyright notice and this permission notice shall be included in
#  all copies or substantial portions of the Software.
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.

def calc(conv):
    def dec(cls):
        funcs = ['__add__','__sub__','__mul__','__radd__','__rsub__','__rmul__','__floordiv__',
        '__truediv__','__mod__']
        for f in funcs:
            def func(fff):
                def gen(self, other):
                    ret = getattr(conv(self),fff)(other)
                    return cls(ret)
                return gen
            setattr(cls, f, func(f))
        return cls
    return dec

def dm(conv):
    def dec(cls):
        funcs = ['__divmod__']
        for f in funcs:
            def func(fff):
                def gen(self, other):
                    mod, m2 = getattr(conv(self),fff)(other)
                    return (cls(mod),cls(m2))
                return gen
            setattr(cls, f, func(f))
        return cls
    return dec

def abpos(conv):
    def dec(cls):
        funcs = ['__neg__','__pos__','__abs__']
        for f in funcs:
            def func(f):
                def gen(self):
                    ret = getattr(conv(self),f)()
                    return cls(ret)
                return gen
            setattr(cls, f, func(f))
        return cls
    return dec


NumRep.py
# Copyright (c) 2015 Muhammed Shamil K
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
# The above copyright notice and this permission notice shall be included in
#  all copies or substantial portions of the Software.
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.

"""
This module provides a class to represent the place vaues in a given number, eg:
        >>> a = NumRep(1234456789123)
        >>> a
        NumRep(Crores=123445,Lakhs=67,Thousands=89,Hundreds=1,Tens=2,Ones=3)
        >>> a.crores
        123445
        >>> a.ones
        3


    This class also provides an all() function, which gives you the real value of the given denomination, 
    eg:
        >>> a = NumRep(1234456789123)
        >>> a.all(HUNDREDS)
        12344567891
"""

HUNDREDS = 'hundreds'
THOUSANDS = 'thousands'
TENS = 'tens'
ONES = 'ones'
LAKHS = 'lakhs'
CRORES = 'crores'

def calc(conv):
    def dec(cls):
        funcs = ['__add__','__sub__','__mul__','__radd__','__rsub__','__rmul__','__floordiv__','__truediv__','__mod__','__divmod__']
        for f in funcs:
            def func(fff):
                def gen(self, other):
                    ret = getattr(conv(self),fff)(other)
                    return cls(ret)
                return gen
            setattr(cls, f, func(f))
        return cls
    return dec

def abpos(conv):
    def dec(cls):
        funcs = ['__neg__','__pos__','__abs__']
        for f in funcs:
            def func(f):
                def gen(self):
                    ret = getattr(conv(self),f)()
                    return cls(ret)
                return gen
            setattr(cls, f, func(f))
        return cls
    return dec


@calc(int)
class RepNum(int):
    def GetRep(self):
        return NumRep(self)


@calc(int)
class NumRep(int):
    """
    This class provides a way to represent the place values in a given number, eg:
        >>> a = NumRep(1234456789123)
        >>> a
        NumRep(Crores=123445,Lakhs=67,Thousands=89,Hundreds=1,Tens=2,Ones=3)
        >>> a.crores
        123445
        >>> a.ones
        3
        >>> a.hundreds
        1
        >>> a.tens
        2
        >>> a.lakhs
        67

    And Every integer in the NumRep has a GetRep() method, which returns the Representation of the number itself,
    eg:
        >>> a = NumRep(123456789)
        >>> a
        NumRep(Crores=12,Lakhs=34,Thousands=56,Hundreds=7,Tens=8,Ones=9)
        >>> a.crores
        12
        >>> a.crores.GetRep()
        NumRep(Tens=1,Ones=2)
        >>> a.crores.GetRep().tens.GetRep()
        NumRep(Ones=1)

    You can compare and do arithmetic operations NumReps with integers and also with other NumReps, like:
        >>> c = NumRep(15)
        >>> c
        NumRep(Tens=1,Ones=5)
        >>> c + 10
        NumRep(Tens=2,Ones=5)
        >>> b = NumRep(10)
        >>> b
        NumRep(Tens=1)
        >>> c + b
        NumRep(Tens=2,Ones=5)


    You can also do arithmetic operations on them and then use GetRep()
    eg:
        >>> a = NumRep(123456789)
        >>> a
        NumRep(Crores=12,Lakhs=34,Thousands=56,Hundreds=7,Tens=8,Ones=9)
        >>> (a.crores - 2).GetRep()
        NumRep(Tens=1)


    This class also provides an all() function, which gives you the real value of the given denomination, 
    eg:
        >>> a = NumRep(1234456789123)
        >>> a.all('hundreds')
        12344567891
        >>> a.all(HUNDREDS)
        12344567891


    """
    __lens = lens = dict(
    lakhs=5,
    crores =7,
    thousands = 3,
    hundreds = 2,
    tens = 1,
    ones = 9,
    )

    def __init__(self, num):
        try:
            num = RepNum(num)
        except TypeError:
            raise
        self.number = RepNum(num)
        self.crores = RepNum(0)
        self.lakhs = RepNum(0)
        self.thousands = RepNum(0)
        self.hundreds = RepNum(0)
        self.tens = RepNum(0)
        self.ones = RepNum(0)
        for time in range(1,9):
            mod = num % 10
            if time == 1:
                self.ones = RepNum(mod)
            if time == 2:
                self.tens = RepNum(mod)
            if time == 3:
                self.hundreds = RepNum(mod)
            if time == 4:
                thou = RepNum(mod)
            if time == 5:
                self.thousands = RepNum(thou + (mod*10))
            if time == 6:
                lak = RepNum(mod)
            if time == 7:
                self.lakhs = RepNum(lak + (mod*10))
            if time == 8:
                self.crores = RepNum(num)
            num -= mod
            num = num/10
        li = []
        if self.crores:
            li.append("Crores={0}".format(self.crores))
        if self.lakhs:
            li.append("Lakhs={0}".format(self.lakhs))
        if self.thousands:
            li.append("Thousands={0}".format(self.thousands))
        if self.hundreds:
            li.append("Hundreds={0}".format(self.hundreds))
        if self.tens:
            li.append("Tens={0}".format(self.tens))
        if self.ones:
            li.append("Ones={0}".format(self.ones))
        self.__li = li
        super().__init__()
    def all(self, digit):
        """
        This function returns the original value of the specified denomination in the number specified,
        not the value in the place.
        eg:
        >>> a = NumRep(1234456789123)
        >>> a.all('hundreds')
        12344567891
        >>> a.all('tens')
        123445678912
        >>> a.all('lakhs')
        12344567
        >>> a.all('crores')
        123445
        >>> a.all(CRORES)
        123445
        
        etc.

        """
        gth = self.__lens.get(digit, None)
        if not gth:
            raise TypeError("Not a valid denomination")
        sr = str(self.number)
        if gth == 9:
            return RepNum(sr)
        return RepNum(sr[:(-gth)])

    def __repr__(self):
        return "NumRep({0})".format(",".join(self.__li))

if __name__ == '__main__':
    import sys
    PY2 = sys.version[0] == "2"
    try:
        while True:
            num = None
            try:
                if PY2:
                    num = int(raw_input("Type the number you want to convert: "))
                else:
                    num = int(input("Type the number you want to convert: "))
            except ValueError:
                print('Not a valid Number')
                continue
            if num:
                print(NumRep(num))
            else:
                sys.exit(0)
    except (KeyboardInterrupt, EOFError):
        sys.exit(0)

